cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java

```diff
@@ -70,99 +74,127 @@ private MetadataEntity(List<KeyValue> details) {
+    metadataEntity.details.put(key, value);
+    metadataEntity.type = key;
+    return metadataEntity;
   }
 
-  @Override
-  public boolean equals(Object o) {
-    if (this == o) {
-      return true;
-    }
-    if (o == null || getClass() != o.getClass()) {
-      return false;
+  /**
+   * Creates a new MetadataEntity with the given type and exists {@link KeyValue} pairs
+   *
+   * @param newType the new type
+   * @return the MetadataEntity which is of the specified type
+   */
+  public MetadataEntity changeType(String newType) {
+    if (newType == null || newType.isEmpty()) {
+      throw new IllegalArgumentException("A valid type must be specified.");
     }
-    MetadataEntity that = (MetadataEntity) o;
-    return Objects.equals(details, that.details);
+    MetadataEntity metadataEntity = new MetadataEntity(this);
+    metadataEntity.type = newType;
+    return metadataEntity;
   }
 
-  @Override
-  public int hashCode() {
-    return details.hashCode();
+  /**
+   * @return the type of the MetadataEntity
+   */
+  public String getType() {
+    return type;
   }
 
-  @Override
-  public String toString() {
-    return "MetadataEntity{" +
-      "details=" + details +
-      '}';
+  /**
+   * @return the value for the given key; if the key does not exists returns null;
+   */
+  @Nullable
+  public String getValue(String key) {
+    return details.get(key);
   }
 
   /**
-   * @return A {@link List} of {@link KeyValue} representing the metadata entity
+   * @return true if there is an entry for the key in the MetadataEntity else false
    */
-  public List<KeyValue> getKeyValues() {
-    return details;
+  public boolean containsKey(String key) {
+    return details.containsKey(key);
   }
 
   /**
-   * @return the value for the key if the key is found else null
+   * @return all the values in the MetadataEntity
    */
-  public String getValue(String key) {
-    if (key == null) {
-      throw new NullPointerException("Key cannot be null");
-    }
-    for (KeyValue detail : details) {
-      if (detail.getKey().equals(key)) {
-        return detail.getValue();
-      }
-    }
-    return null;
+  public Iterable<String> getValues() {
+    return Collections.unmodifiableList(new ArrayList<>(details.values()));
   }
 
   /**
-   * @return A {@link Iterator} over the keys in this {@link MetadataEntity}
+   * @return all the keys in the MetadataEntity
    */
   public Iterable<String> getKeys() {
-    return details.stream().map(KeyValue::getKey)::iterator;
+    return Collections.unmodifiableList(new ArrayList<>(details.keySet()));
   }
 
   /**
-   * @return A {@link Iterator} over the values in this {@link MetadataEntity}
+   * @return A {@link List} of {@link KeyValue} representing the metadata entity
    */
-  public Iterable<String> getValues() {
-    return details.stream().map(KeyValue::getValue)::iterator;
+  @Override
+  public Iterator<KeyValue> iterator() {
+    List<KeyValue> result = new LinkedList<>();
```

Comment:
could you use Iterables.transform(details.entrySet(), entry -> new KeyValue(entry.getKey(), getValue()).iterator?